### PR TITLE
Always display python type information in cast errors

### PR DESCRIPTION
**Pull request by virtuald**
## Description

I opened a discussion at https://github.com/pybind/pybind11/discussions/4458:

---

Currently, the error message is:

```
RuntimeError: Unable to cast Python instance to C++ type (#define PYBIND11_DETAILED_ERROR_MESSAGES or compile in debug mode for details)
```

That's not useful. We should add the python type to this error message to help users out even though we don't have the C++ type. This PR does that.

## Suggested changelog entry:

N/A

### CODE REVIEW

File: include/pybind11/cast.h [hunk 1]
```diff
@@ -1017,8 +1017,10 @@ type_caster<T, SFINAE> &load_type(type_caster<T, SFINAE> &conv, const handle &ha
                   "Internal error: type_caster should only be used for C++ types");
     if (!conv.load(handle, true)) {
 #if !defined(PYBIND11_DETAILED_ERROR_MESSAGES)
-        throw cast_error("Unable to cast Python instance to C++ type (#define "
-                         "PYBIND11_DETAILED_ERROR_MESSAGES or compile in debug mode for details)");
+        throw cast_error(
+            "Unable to cast Python instance of type " + (std::string) str(type::handle_of(handle))
```
Comment (Skylion007) [inline include/pybind11/cast.h:1021]: Can we avoid a c style class here and make it a static one instead?

Comment (virtuald) [reply include/pybind11/cast.h:1021]: Sure. I was just copying the code block below it.

**Comment by rwgk** [comment]
I'm unsure about this change. My guess: the original idea was to keep the release binaries small, and make the debug behavior comprehensive. This PR waters down the motivation for having the release/debug difference. Questions that cross my mind:

* Do we want to accept the binary size increases?
* If we do, why not go all the way to the comprehensive messages also for release builds? That would be a nice simplification.

To make that decision scientifically, we'd need some measurements, for typical situations.

Without having data: I feel it's best to either not touch this, or get rid of the release/debug difference altogether.

**Comment by virtuald** [comment]
I don't know the answer to your questions, but:

* `RuntimeError: Unable to cast Python instance to C++ type` is a useless error message. Generally python tries to give good error messages, so should pybind11. 
* It always seemed to me (without actual knowledge) that the rationale for the debug/release split was to avoid bringing in the C++ type information. The python type information is already there, so it doesn't seem like this would add very much.

**Comment by virtuald** [comment]
I picked one of my packages that has a bunch of manual pybind11 code and typical class/function wrappers. It uses smart_holder, but here are some non-scientific results from that exploration:

## [pyntcore](https://github.com/robotpy/pyntcore), Linux gcc 12.2.1, std=c++20, release build

| What | Size | %diff | Stripped Size |%diff stripped |
|-|-|-|-|-|
| smart-holder 8ece7d641ca6ce316e59fec6744b8517073bbe32 | 4318976 | 0% | 2853288 |  0% |
| display-python-type-smart-holder 129fc50d0704b9a4cacbd4012e2f81e42e1ed121 | 4352464 | +0.7% |2886056 |  +1.5% | 
|  smart-holder 8ece7d641ca6ce316e59fec6744b8517073bbe32 and -DPYBIND11_DETAILED_ERROR_MESSAGES=1 | 4471904 | +3.5% |3000744 |  +5.1%

**Comment by rwgk** [comment]
> here are some non-scientific results from that exploration

Scientific enough for me! That sways my opinion:
* +1.5% for a much more useful error message is a good bargain.
* The delta to +5.1% is significant enough to still keep the debug/release distinction (IOW the `PYBIND11_DETAILED_ERROR_MESSAGES` `#ifdefs`).

I'm assuming the measurements are similar on other platforms. Seems like a pretty safe assumption.

**Review by rwgk — COMMENTED** [review]
Two more suggestions/questions, basically retro-fitting, to improve alignment with best practices:

* Change C-style cast to `static_cast` (or even better IMO `str(...).cast<std::string>()`) not only in the copy-pasted code, but also in the original. To not create weird asymmetries.

* Is all changed code covered by unit tests? I didn't look systematically, but IIUC there are more code changes than test changes. If that's true, now is the right time to plug those holes.

Small worry: this is a behavior change that could break test code. After your pass through my suggestions I'll run global testing to see if that's actually something to worry about.

File: include/pybind11/detail/common.h [hunk 2]
```diff
@@ -1227,6 +1227,11 @@ constexpr
 // Pybind offers detailed error messages by default for all builts that are debug (through the
 // negation of ndebug). This can also be manually enabled by users, for any builds, through
 // defining PYBIND11_DETAILED_ERROR_MESSAGES.
+//
```
[the remark this answers is not in the document]
Comment (virtuald) [reply include/pybind11/detail/common.h:1230]: It doesn't just add C++ type information. It also adds some GIL checks (gil.h) and some extra `py::arg` related checks (attr.h).

**Comment by virtuald** [comment]
The other cases seem to have test coverage, they were just generic and didn't check the whole exception.

**Review by rwgk — APPROVED** [review]
LGTM for code changes.

Looking for another review from @Skylion007 or @henryiii

**Comment by rwgk** [comment]
I ran Google global testing for this PR and it passed (internal test ID OCL:504179773:BASE:504260517:1674571457025:b95799a2).

**Review by Skylion007 — APPROVED** [review]
Really tiny nit, but some of these strings could be constructed using std::string("...").append() other strings for a bit more efficiency: https://clang.llvm.org/extra/clang-tidy/checks/performance-inefficient-string-concatenation.html. This looks good either way though.

**Comment by virtuald** [comment]
Comment updated.

**Comment by rwgk** [comment]
Thanks Dustin, I'll merge this now.

Regarding GHA, the Pip failure is expected (currently unresolved: PR #4497), the CI / 🐍 pypy-3.9 • windows-2022 • x64 failure is a long-standing flake.

**Comment by rwgk** [comment]
I added a suggested changelog entry. Feel free to tweak.

**Pull request merged**